### PR TITLE
Refactor backup state handling in `backup_monitor.go` and `backup.go`

### PR DESCRIPTION
**Pull request by derekbit**
- Change the return type and add variable initialization in `syncBackupStatusFromEngineReplica` in `backup_monitor.go`
- Add error handling and update `currentBackupStatus` in `syncBackupStatusFromEngineReplica` in `backup_monitor.go`
- Add a new constant `backupStateStarting` in `backups. Go`
- Add a new backup state `Starting` in `backup. Go`

Depends on https://github.com/longhorn/backupstore/pull/117 and https://github.com/longhorn/longhorn-engine/pull/882
Longhorn/longhorn#5662

### CODE REVIEW

**Comment by innobead** [comment]
@derekbit need to fix the conflict.

File: engineapi/backups.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 )
 
 const (
+	backupStateStarting   = "starting"
```
Comment (innobead) [inline engineapi/backups.go:23]: Can we just use these constants from https://github.com/longhorn/backupstore/commit/cc70e06e?

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM.

However, please consolidate commits since some commit seems irrelevent like CRD update for share manager resource. 

After engine PR merged and the go dependencies updated, we can merge this PR.

File: engineapi/backups.go (hunk 1 above)
Comment (derekbit) [reply engineapi/backups.go:23]: The introduction of `starting` state makes the incompatibility of the new `longhorn-manager` and the old `longhorn-engine`. So, checking the engine CLI/API version is necessary. Do we need to include the feature in v1.4.x? 
cc @innobead

Comment (innobead) [reply engineapi/backups.go:23]: Where is the incompatible part? I saw all changes are internally for backupstore lib.

Comment (innobead) [reply engineapi/backups.go:23]: For any incompatible change, as discussed before, no need, and be unsuitable for backport. 

cc @longhorn/dev

**Comment by derekbit** [comment]
The purpose of the PR is to handle the side effect of `soft` mode.
In addition, the introduction of the `starting` state will break the compatibility of the new longhorn-manager and the old backupstore. So, bumping cli/api version and version check before doing a backup is necessary.
Should we backport `soft` mode to v1.4.2? Ref: https://github.com/longhorn/longhorn/issues/5628

cc @innobead

File: engineapi/backups.go (hunk 1 above)
Comment (derekbit) [reply engineapi/backups.go:23]: https://github.com/longhorn/longhorn-manager/pull/1813#issuecomment-1531327304

**Comment by innobead** [comment]
> The purpose of the PR is to handle the side effect of `soft` mode. In addition, the introduction of the `starting` state will break the compatibility of the new longhorn-manager and the old backupstore. So, bumping cli/api version and version check before doing a backup is necessary. Should we backport `soft` mode to v1.4.2? Ref: [longhorn/longhorn#5628](https://github.com/longhorn/longhorn/issues/5628)
> 
> cc @innobead

Then, we don't need to backport any incompatible change/fix.

**Comment by derekbit** [comment]
> > The purpose of the PR is to handle the side effect of `soft` mode. In addition, the introduction of the `starting` state will break the compatibility of the new longhorn-manager and the old backupstore. So, bumping cli/api version and version check before doing a backup is necessary. Should we backport `soft` mode to v1.4.2? Ref: [longhorn/longhorn#5628](https://github.com/longhorn/longhorn/issues/5628)
> > cc @innobead
> 
> Then, we don't need to backport any incompatible change/fix.

OK, I will revert the `soft` mode codes in https://github.com/longhorn/backupstore/commits/v1.4.x

**Comment by innobead** [comment]
Thanks @derekbit !

**Comment by innobead** [comment]
@derekbit before reverting, let see the severity first.

What issue the soft mod is trying to fix? IIRC, it's about High cpu loading due to stuck nfs connection?

File: k8s/pkg/apis/longhorn/v1beta2/backup.go [hunk 2]
```diff
@@ -7,6 +7,7 @@ type BackupState string
 const (
 	BackupStateNew        = BackupState("")
 	BackupStatePending    = BackupState("Pending")
+	BackupStateStarting   = BackupState("Starting")
```
Comment (innobead) [inline k8s/pkg/apis/longhorn/v1beta2/backup.go:10]: This is not necessary, since we did not introduce this state from backup status anymore.

**Review by innobead — CHANGES_REQUESTED** [review]

File: k8s/pkg/apis/longhorn/v1beta2/backup.go (hunk 2 above)
Comment (derekbit) [reply k8s/pkg/apis/longhorn/v1beta2/backup.go]: Removed. Thank you.

**Pull request merged**